### ISSUE TITLE
Accuracy issue with CustomDecoder

### ISSUE THREAD

**Issue by Sanster** [NVIDIA/FasterTransformer#154]
- FasterTransformer: v4.0
- GPU: RTX3090
- Driver: 460.91.03
- CUDA: 11.1
- python: 3.6.9
- OpenNMT-py: 1.1.1
- pytorch vesion: 1.8.2+cu111
- Build with sm86 according to this issue https://github.com/NVIDIA/FasterTransformer/issues/90

Running `decoder_sample.py` in [decoder_guide](https://github.com/NVIDIA/FasterTransformer/blob/main/docs/decoder_guide.md#decoder-and-decoding), the output differences between FT and opennmt-py is very large.

```bash
./bin/decoding_gemm 8 4 8 64 31538 32 512 0
python pytorch/decoder_sample.py 8 6 32 8 64
```

![image](https://user-images.githubusercontent.com/3998421/136642716-6de4bc24-a2bb-4a62-b209-c9e439ab9eef.png)

The output of `encoder_sample.py` is normal
<img width="331" alt="截屏2021-10-09 上午11 46 39" src="https://user-images.githubusercontent.com/3998421/136643193-408db592-3331-494f-9448-49864a9a35e9.png">

**Comment by byshiue** [comment]
Try to set
```bash
export NVIDIA_TF32_OVERRIDE=0
```
to disable the TF32.

**Comment by Sanster** [comment]
> Try to set
> 
> ```shell
> export NVIDIA_TF32_OVERRIDE=0
> ```
> 
> to disable the TF32.

Thanks, it works on 3090
![image](https://user-images.githubusercontent.com/3998421/136653862-004e64e1-974a-4d1d-aa05-4e475d106070.png)

===============

For those who may be interested in [MASTER-pytorch](https://github.com/wenwenyu/MASTER-pytorch), I am working on convert [Decoder part](https://github.com/wenwenyu/MASTER-pytorch/blob/main/model/transformer.py#L176) to FT。I will share the results here, currently the predictions are not right yet.

pytorch result:

![image](https://user-images.githubusercontent.com/3998421/136653989-7bef938e-b6c2-4c80-9718-d3de3480467a.png)


FT Result:

![image](https://user-images.githubusercontent.com/3998421/136653882-7cfaf687-4f0b-4221-b063-7adef5b3bbf8.png)

**Comment by byshiue** [comment]
1. Check that you put the pytorch weights into FT correctly. 
2. You can compare the intermediate results between MASTER-pytorch and opennmt first because both of them are on pytorch. Guarantee that they have same model architecture.

**Comment by Sanster** [comment]
I created a working example here: https://github.com/Sanster/FasterTransformer/tree/v4.0-master-example/sample/pytorch/master_example

I found that if `ffn_hidden_size != 4 * hidden_dim(head_num*head_size)`, the result is wrong. **Does the hidden size of FFN in decoder have to be equal to `4 * hidden_dim(head_num*head_size)`?**

Result of `ffn_hidden_size == 4 * hidden_dim(head_num*head_size)`

```
  step  diff name      max relative diff
------  -----------  -------------------
     0  ft-master            7.16644e-05
     0  ft-onmt              4.15602e-05
     0  onmt-master          3.26929e-05
     1  ft-master            0.000143105
     1  ft-onmt              0.000181676
     1  onmt-master          3.85782e-05
     2  ft-master            0.000124314
     2  ft-onmt              0.000559099
     2  onmt-master          0.00060457
     3  ft-master            9.97579e-05
     3  ft-onmt              0.000107287
     3  onmt-master          6.91183e-05
```


Result of `ffn_hidden_size != 4 * hidden_dim(head_num*head_size)`
```
  step  diff name      max relative diff
------  -----------  -------------------
     0  ft-master           53.7376
     0  ft-onmt             53.7376
     0  onmt-master          0.000292255
     1  ft-master          103.101
     1  ft-onmt            103.101
     1  onmt-master          0.000151085
     2  ft-master          209.367
     2  ft-onmt            209.367
     2  onmt-master          0.000504965
     3  ft-master          145.746
     3  ft-onmt            145.746
     3  onmt-master          0.00112149
```

**Comment by Sanster** [comment]
🥲 I didn't find `ffn_hidden_size is fixed to 4 * hidden_dim` in the documentation. Maybe it would be better to add a note in the documentation

**Comment by 520jefferson** [comment]
> In 4.0, the ffn_hidden_size is fixed to 4 * hidden_dim. We relax the constraint in v5.0_beta, you can refer https://github.com/Sanster/FasterTransformer/blob/dev/v5.0_beta/examples/tensorflow/decoder/decoder_example.py and set by `inter_size`. But we still may change the API recently.

I also met this error. ffn_hidden_size = 3072 and hidden_dim=768 .Same config， FT2.0
